macros/one_hot_encoder.sql

```diff
@@ -1,56 +1,89 @@
+            {% set error_message %}
+        The `one_hot_encoder` macro only supports one of include_columns and exclude_columns being a non-default value or one of them being a list.
+            {% endset %}
+            {%- do exceptions.raise_compiler_error(error_message) -%}
+        {%- elif include_columns!='*' and (include_columns is not iterable  or include_columns is string or include_columns is mapping) -%}
+            {% set error_message %}
+        The `one_hot_encoder` macro only supports one of include_columns and exclude_columns being a non-default value or one of them being a list.
+            {% endset %}
+            {%- do exceptions.raise_compiler_error(error_message) -%}
+        {%- elif exclude_columns is not none and (exclude_columns is not iterable or exclude_columns is string or exclude_columns is mapping) -%}
+                    {% set error_message %}
+        The `one_hot_encoder` macro only supports one of include_columns and exclude_columns being a non-default value or one of them being a list.
+            {% endset %}
+            {%- do exceptions.raise_compiler_error(error_message) -%}
+        {%- endif -%}
+    {%- endif -%}
+    {{ adapter.dispatch('one_hot_encoder',packages=['dbt_ml_preprocessing'])(source_table, source_column, category_values, handle_unknown, include_columns, exclude_columns) }}
 {%- endmacro %}
 
-{% macro default__one_hot_encoder(source_table,source_column,category_values,handle_unknown,include_columns) %}
-select 
-{% for column in include_columns %}
-{{ column }},
-{% endfor %}
-{% for category in category_values %}
-case {{source_column}}
-    when '{{category}}' then true
-    else false
-    end as {{source_column}}_{{category}}
-{% if not loop.last %}, {% endif %}
-{% endfor %}
-from {{ source_table }}
+{% macro default__one_hot_encoder(source_table, source_column, category_values, handle_unknown, include_columns, exclude_columns) %}
+    {% set columns = adapter.get_columns_in_relation( source_table ) %}
+
+
+
+
+    with binary_output as (
+    select
+        {%- if include_columns=='*' and exclude_columns is  none -%}
+            {% for column in columns %}
+                {{ column.name }},
+            {%- endfor -%}
+        {%- elif include_columns !='*'-%}
+            {% for column in include_columns %}
+                {{ source_table }}.{{ column }},
+            {%- endfor -%}
+        {%- else -%}
+            {% for column in columns %}
+            {%- if column.name | lower not in exclude_columns | lower %}
+                {{ column.name }},
+            {%- endif -%}
+            {%- endfor -%}
+        {%- endif -%}
+        {% for category in category_values %}
+            {% set no_whitespace_column_name = category | replace( " ", "_") -%}
+                {%- if category is none %}
+                    case when {{ source_column }} is null then 1 else 0 end as is_{{ source_column }}_null
```

Comment:
In the scikit-learn implementation (at least as of sometime last year), the behaviour is governed here:

>handle_unknown{‘error’, ‘ignore’}, default=’error’
>Whether to raise an error or ignore if an unknown categorical feature is present during transform (default is to raise). When this parameter is set to ‘ignore’ and an unknown category is encountered during transform, the resulting one-hot encoded columns for this feature will be all zeros. In the inverse transform, an unknown category will be denoted as None.

(see https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html)

So instead of providing a is_{{ source_column }}_null column, I decided to implement the 'handle_unknown' parameter value 'error', to mirror scikit-learn behaviour.

This means users must either handle missing values upstream, or accept that they create 'false' outputs in all columns.

I've made this change and it's all working, just a bit stuck on getting a test case for this: https://getdbt.slack.com/archives/C2JRRQDTL/p1614556090266800